Prototype update [1.21]

LizzieSpace opened the pull request:
- Update Java version and dependencies

Updated the Java version to 21 across all workflows and in the build files. Dependencies including the Shadow plugin and Architectury Loom have also been updated to their latest versions. Some changes were made to codes to support the new Java version.

_I'm not especially experienced in mod making, so it's probably better to call this a sort of beta test. I shall be conducting further stress tests to see if my update is stable._

I've forked and updated the mod to 1.21 with solely `FABRIC` in mind in order to use it in a private ongoing project. Further progress in the update, such as patches and whatever else, will be updated through this same channel.

LizzieSpace committed:
Updated to 1.21

- Update Java version and dependencies

Updated the Java version to 21 across all workflows and in the build files. Dependencies including the Shadow plugin and Architectury Loom have also been updated to their latest versions. Some changes were made to codes to support the new Java version.

Signed-off-by: Liz C. <alicecfire@gmail.com>

LizzieSpace commented:
I'll be reviewing why the build failed as soon as I'm free from my uni studies.

It was working on my side last I messed around with the code so...

I guess that's what I mean by WIP.

samolego commented:
It's failing to fetch that confog2brigadier version.

LizzieSpace commented:
I mean yeah I can gather as much, but I don't quite understand why my machine can get the version alright while the pipeline can't.

Due to time constraints, I haven't had the chance to review it further, but an associate of mine suggested it may be an access issue.

I'll be checking it out later tonight…


---
Update: I ran the same workflows on my fork, only changing the workflows' java version to 21, and it worked…

[![Fabric Build](https://github.com/LizzieSpace/ChestRefill/actions/workflows/fabric-build.yml/badge.svg)](https://github.com/LizzieSpace/ChestRefill/actions/workflows/fabric-build.yml)

LizzieSpace committed:
minor stuff

Signed-off-by: Liz C. <alicecfire@gmail.com>

enderkruemel commented:
is there any working version for 1.21 aviable? i cant get it to work if i compiled it myselve...
btw i never used gradlew or anything like this before, i only used to write datapacks.

enderkruemel commented:
a friend helpt me to compile it , after removing all forge stuff it worked just fine.

LizzieSpace commented:
Well... To be very fair I did only mess with the fabric part... I don't know what I expected really 

@enderkruemel do be warned that it's a very rough update and I only looked far enough into it to make the fabric side work. working and working as intended are different things. 

I haven't thoroughly tested this fork. I wouldn't say you should expect errors but sure should prepare for if one does happen.

---

Now that you gave me a good enough tip that I should stop goofing around and work on the forge side of things, I'll work on actually preparing this for a merge once I've got time on my tiny raccoon hands.

Thanks for your feedback!

enderkruemel commented:
i only needet the fabric version so, thank you for the 1.21 port

LizzieSpace commented:
@enderkruemel if you find any problems directly related to my version, feel free to make a bug report at the fork and I'll investigate

enderkruemel commented:
ok if i find any problem i will do this 👍

enderkruemel commented:
@LizzieSpace @samolego would it be possible to add the function to reloot a chest when an other player looted the chest? 
i think about a setting to allow reloot but only afer someone elese loots the chest.
so i loot, a friend comes some time later loots again, after this an other one comes and loots, and than i can loot the same chest again. it would make this a lot more usefull i think .

LizzieSpace commented:
I'm pretty sure that should already be a feature tho... I'll look into it

@enderkruemel does the `allowReloot` tag not serve you?

enderkruemel commented:
@LizzieSpace not realy,  it makes so you can reloot every time it refills, not only when someone else used  the same chest after you...

samolego commented:
@enderkruemel sorry, this is not the right place. Open an issue an discuss it there, this PR is for different stuff :)

LizzieSpace commented:
@samolego I'll finalize the work on my PR and check the issue

LizzieSpace committed:
build: update project configuration and dependencies

This commit makes several updates to the project configuration files and dependencies.

- The "forge" module has been commented out in the settings.gradle file, as it is a work in progress.
- The yarn mappings have been replaced with parchment mappings in gradle.properties.
- The Architectury plugin configuration in common/build.gradle has been updated to include both "fabric" and "forge", as the previous implementation was deprecated.
- The Architectury Loom has been updated to V1.7-SNAPSHOT
- A new Maven repository for ParchmentMC has been added to build.gradle.

LizzieSpace committed:
ci: update GitHub Actions to use latest action versions

This update modifies the GitHub Actions workflows to utilize the latest versions of various actions. Specifically, the `actions/cache`, `actions/checkout`, `gradle/wrapper-validation-action`, and `actions/upload-artifact` actions have been upgraded from version 2 or 1 to version 4 or 3 respectively. These changes ensure improved performance, security, and compatibility with current standards.

- Updated workflow files include: forge-release.yml, fabric-build.yml, fabric-release.yml, and forge-build.yml.
- The changes aim to leverage enhancements provided by the newer versions of the actions.
- This upgrade may also address any deprecation warnings or bugs in the previous action versions.

LizzieSpace committed:
fix: update loader version and improve mixin code structure

This commit updates the loader version in the gradle properties and enhances the structure of the RandomizableContainerBEMixin_LootRefiller class. The changes include refactoring the code for better readability and maintainability by removing unnecessary method definitions and optimizing imports. Additionally, the icon path in the fabric.mod.json file has fixed.

- Updated loader version from 0.15.11 to 0.16.0 for compatibility.
- Refactored mixin methods for clarity and removed unused code.
- fixed icon path in fabric.mod.json.

LizzieSpace commented on common/build.gradle:
```diff
@@ -6,7 +6,7 @@ dependencies {
 
 architectury {
-    common()
+    common("fabric","forge")
```
The previous usage of `common()` was deprecated.

LizzieSpace commented on build.gradle:
```diff
@@ -1,6 +1,6 @@
 plugins {
     id "architectury-plugin" version "3.4-SNAPSHOT"
-    id "dev.architectury.loom" version "1.0-SNAPSHOT" apply false
+    id "dev.architectury.loom" version "1.7-SNAPSHOT" apply false
```
Updated Architectury loom to its latest version.

LizzieSpace commented on common/src/main/java/org/samo_lego/chestrefill/mixin/RandomizableContainerBEMixin_LootRefiller.java:
```diff
@@ -83,25 +115,49 @@ private void refillLootTable(@Nullable Player player, CallbackInfo ci) {
+     * @param compoundTag The compound tag containing the loot table information.
+     * @return true if the loot table was successfully loaded and set, false otherwise.
+     */
+    public boolean tryLoadLootTable(@NotNull CompoundTag compoundTag) {
```
This segments overwrites the default method `tryLoadLootTable`, as it is no longer possible to inject code onto it due to it being now a default function on `RandomizableContainer`.

LizzieSpace commented on common/src/main/java/org/samo_lego/chestrefill/mixin/RandomizableContainerBEMixin_LootRefiller.java:
```diff
@@ -83,25 +115,49 @@ private void refillLootTable(@Nullable Player player, CallbackInfo ci) {
     }
 
-    @Inject(method = "tryLoadLootTable", at = @At("RETURN"))
-    private void onLootTableLoad(CompoundTag compoundTag, CallbackInfoReturnable<Boolean> cir) {
```
this no longer works as of 1.21

LizzieSpace commented on common/src/main/java/org/samo_lego/chestrefill/mixin/RandomizableContainerBEMixin_LootRefiller.java:
```diff
@@ -138,8 +194,24 @@ private void onLootTableLoad(CompoundTag compoundTag, CallbackInfoReturnable<Boo
+    public boolean trySaveLootTable(CompoundTag compoundTag) {
+        this.onLootTableSave(compoundTag);
+        return RandomizableContainer.super.trySaveLootTable(compoundTag);
+    }
```
Instead of injecting on the HEAD a method that no longer exists on the target class, we simply implement it as per the `RandomizableContainer` interface, calling its default implementation once we're done

LizzieSpace commented on fabric/src/main/resources/fabric.mod.json:
```diff
@@ -22,7 +22,7 @@
   "mixins": ["chestrefill.common.mixins.json"],
   "depends": {
-    "fabricloader": ">=0.7.4",
+    "fabricloader": ">=0.16.0",
```
Due to https://github.com/SpongePowered/Mixin/issues/504, Previous fabric Loader versions are incompatible with the rewritten LootRefiller Implementation.

LizzieSpace commented on settings.gradle:
```diff
@@ -10,6 +10,6 @@ pluginManagement {
 include("common")
 include("fabric")
-include("forge")
+//include("forge")
```
I have tried my best to make the forge side of this project work, however I've been utterly unable to and thus have disabled this project segment

LizzieSpace commented:
@samolego This might be the closest to ready that I can get this pr. I've been unable to figure out the forge segment, but the fabric one should work as intended. (or at least it very much did throughout my tests)

samolego commented:
Hi! Don't worry about forge, it's pain the backside. I've migrated most of my other projects back to basic fabric loom even.

Thanks for your time. I'll take a closer look later today.

LizzieSpace commented:
oh apparently the pr closes if i rename the branch dang...

LizzieSpace commented:
@samolego After further testing i've found that the nbt breaks after reopening the world. it seems it is not being correctly saved. i'm attempting to trace the issue rn and will come back with a resolution asap

samolego commented:
Oof, ok. No problem, haven't tested & released yet.
